vm/virtualbox/virtualbox.go

```diff
@@ -0,0 +1,312 @@
+		"modifyvm", inst.vmName, "--uart-mode1", "server", inst.serialPath); err != nil {
+		if inst.debug {
+			log.Logf(0, "VBoxManage modifyvm --uart-mode1 failed: %v", err)
+		}
+		return err
+	}
+
+	return nil
+}
+
+// VirtualBox waits for a serial link during boot; this opens the socket so the VM can continue.
+func pokeSerialSocket(inst *instance, d time.Duration) error {
+	conn, err := net.Dial("unix", inst.serialPath)
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+
+	done := make(chan struct{})
+	deadline := time.Now().Add(d)
+
+	go func() {
+		defer close(done)
+		buf := make([]byte, 4096)
+
+		for {
+			_ = conn.SetReadDeadline(deadline)
+			if _, err := conn.Read(buf); err != nil {
+				var ne net.Error
+				if errors.As(err, &ne) && ne.Timeout() {
+					return
+				}
+				if !errors.Is(err, io.EOF) && inst.debug {
+					log.Logf(0, "serial read error: %v", err)
+				}
+				return
+			}
+		}
+	}()
+
+	<-done
+	return nil
+}
+
+func (inst *instance) boot() error {
+	if inst.debug {
+		log.Logf(0, "booting VM %q (headless)", inst.vmName)
+	}
+	if _, err := osutil.RunCmd(2*time.Minute, "", "VBoxManage",
+		"startvm", inst.vmName, "--type", "headless"); err != nil {
+		if inst.debug {
+			log.Logf(0, "VBoxManage startvm failed: %v", err)
+		}
+		return err
+	}
+
+	time.Sleep(10 * time.Second)
+	if err := pokeSerialSocket(inst, 30*time.Second); err != nil {
+		if inst.debug {
+			log.Logf(0, "failed to poke serial socket %s: %v", inst.serialPath, err)
+		}
+	}
+	return nil
+}
+
+func (inst *instance) Forward(port int) (string, error) {
+	if inst.rpcPort != 0 {
+		return "", fmt.Errorf("isolated: Forward port already set")
+	}
+	if port == 0 {
+		return "", fmt.Errorf("isolated: Forward port is zero")
+	}
+	inst.rpcPort = port
+	return fmt.Sprintf("127.0.0.1:%d", port), nil
+}
+
+func (inst *instance) Close() error {
+	if inst.debug {
+		log.Logf(0, "stopping %v", inst.vmName)
+	}
+	osutil.RunCmd(2*time.Minute, "", "VBoxManage", "controlvm", inst.vmName, "poweroff")
+	if inst.debug {
+		log.Logf(0, "deleting %v", inst.vmName)
+	}
+	osutil.RunCmd(2*time.Minute, "", "VBoxManage", "unregistervm", inst.vmName, "--delete")
+	close(inst.closed)
+	return nil
+}
+
+func (inst *instance) Copy(hostSrc string) (string, error) {
+	base := filepath.Base(hostSrc)
+	vmDest := "/" + base
+
+	args := vmimpl.SCPArgs(inst.debug, inst.sshkey, inst.sshPort, false)
+	args = append(args, hostSrc, fmt.Sprintf("%v@127.0.0.1:%v", inst.sshuser, vmDest))
+
+	if inst.debug {
+		log.Logf(0, "running command: scp %#v", args)
+	}
+
+	if _, err := osutil.RunCmd(3*time.Minute, "", "scp", args...); err != nil {
+		return "", err
+	}
+	return vmDest, nil
+}
+
+func (inst *instance) Run(ctx context.Context, command string) (
+	<-chan []byte, <-chan error, error) {
+	dmesg, err := net.Dial("unix", inst.serialPath)
+	if err != nil {
+		if inst.debug {
+			log.Logf(0, "serial console not available: %v; continuing without it", err)
```

Comment:
Nit: is this ever an expected scenario?
Just returning with an error here would simplify the rest of the code.